apm/monitor_test.go

```diff
@@ -0,0 +1,241 @@
+package apm
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/evergreen-ci/birch"
+	"github.com/mongodb/ftdc"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/event"
+)
+
+func TestMonitor(t *testing.T) {
+	m, ok := NewBasicMonitor(nil).(*basicMonitor)
+	require.True(t, ok)
+	t.Run("Tracking", func(t *testing.T) {
+		t.Run("Pop", func(t *testing.T) {
+			t.Run("Empty", func(t *testing.T) {
+				assert.Len(t, m.inProg, 0)
+				key := m.popRequest(42)
+				assert.Zero(t, key)
+				assert.Len(t, m.inProg, 0)
+			})
+			t.Run("Existing", func(t *testing.T) {
+				m.inProg[42] = eventKey{dbName: "amboy"}
+				key := m.popRequest(42)
+				assert.Equal(t, "amboy", key.dbName)
+				assert.Len(t, m.inProg, 0)
+			})
+		})
+		t.Run("Set", func(t *testing.T) {
+			resetMonitor(t, m)
+			t.Run("WithValue", func(t *testing.T) {
+				m.setRequest(42, eventKey{cmdName: "find"})
+				assert.Len(t, m.inProg, 1)
+				k := m.popRequest(42)
+				assert.Equal(t, "find", k.cmdName)
+			})
+			t.Run("Filter", func(t *testing.T) {
+				m.config = &MonitorConfig{
+					Databases: []string{"amboy"},
+					Commands:  []string{"find"},
+				}
+				assert.Len(t, m.inProg, 0)
+				m.setRequest(42, eventKey{cmdName: "find"})
+				assert.Len(t, m.inProg, 0)
+				m.setRequest(42, eventKey{dbName: "amboy", cmdName: "find"})
+				assert.Len(t, m.inProg, 1)
+				m.config = nil
+			})
+		})
+		t.Run("Get", func(t *testing.T) {
+			resetMonitor(t, m)
+			t.Run("Empty", func(t *testing.T) {
+				r := m.getRecord(42)
+				require.Nil(t, r)
+
+				assert.Len(t, m.current, 0)
+			})
+			t.Run("Zeroed", func(t *testing.T) {
+				resetMonitor(t, m)
+				m.inProg[42] = eventKey{}
+				r := m.getRecord(42)
+				assert.Len(t, m.inProg, 0)
+				require.Nil(t, r)
+
+				assert.Len(t, m.current, 0)
+			})
+			t.Run("PartialData", func(t *testing.T) {
+				resetMonitor(t, m)
+				m.inProg[42] = eventKey{dbName: "amboy", cmdName: "find"}
+				r := m.getRecord(42)
+				assert.NotNil(t, r)
+				assert.Len(t, m.inProg, 0)
+
+				assert.Len(t, m.current, 1)
+			})
+			t.Run("MultipleData", func(t *testing.T) {
+				resetMonitor(t, m)
+				m.inProg[42] = eventKey{dbName: "amboy", collName: "jobs", cmdName: "find"}
+				r := m.getRecord(42)
+				assert.NotNil(t, r)
+				assert.Len(t, m.inProg, 0)
+
+				assert.Len(t, m.current, 1)
+			})
+		})
+	})
+	t.Run("Collector", func(t *testing.T) {
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		resetMonitor(t, m)
+		collector := m.DriverAPM()
+		require.NotNil(t, collector)
+		t.Run("StartEventFind", func(t *testing.T) {
+			collector.Started(ctx, &event.CommandStartedEvent{
+				DatabaseName: "amboy",
+				CommandName:  "find",
+				RequestID:    42,
+				Command: buildCommand(t, birch.DC.Elements(
+					birch.EC.String("find", "jobs"),
+				)),
+			})
```

Comment:
Shouldn't we check that that the event was added?

Reply:
done